tests/builders/test_descriptors.py

```diff
@@ -0,0 +1,137 @@
+            return iter([x for x in templates if isinstance(x, ParameterTemplate)])
+
+    class FakeProject(Project):
+        def __init__(self):
+            pass
+
+        @property
+        def property_templates(self) -> PropertyTemplateCollection:
+            return FakePropertyTemplateCollection()
+
+        @property
+        def condition_templates(self) -> ConditionTemplateCollection:
+            return FakeConditionTemplateCollection()
+
+        @property
+        def parameter_templates(self) -> ParameterTemplateCollection:
+            return FakeParameterTemplateCollection()
+
+    return FakeProject()
+
+
+def test_valid_template_conversions():
+    expected = [
+        (
+            PropertyTemplate(name="density", bounds=RealBounds(lower_bound=0, upper_bound=100, default_units="g/cm^3")),
+            density_desc
+        ),
+        (
+            ConditionTemplate(name="speed", bounds=CategoricalBounds(categories=["low", "high"])),
+            CategoricalDescriptor(key="speed", categories=["low", "high"])
+        ),
+        (
+            ParameterTemplate(name="solvent", bounds=MolecularStructureBounds()),
+            MolecularStructureDescriptor(key="solvent")
+        ),
+        (
+            PropertyTemplate(name="formula", bounds=CompositionBounds(components=EmpiricalFormula.all_elements())),
+            ChemicalFormulaDescriptor(key="formula")
+        )
+    ]
+
+    for tmpl, desc in expected:
+        assert template_to_descriptor(tmpl) == desc
+
+
+def test_invalid_template_conversions():
+    with pytest.raises(ValueError):
+        template_to_descriptor(
+            ConditionTemplate("foo", bounds=IntegerBounds(lower_bound=0, upper_bound=1))
+        )
+
+    with pytest.raises(ValueError):
+        template_to_descriptor(
+            PropertyTemplate("mixture", bounds=CompositionBounds(components=["sugar", "spice"]))
+        )
+
+    class DummyBounds(BaseBounds):
+        """Fake bounds to test unrecognized bounds."""
+
+        def contains(self, bounds):
+            return False
+
+    with pytest.raises(ValueError):
+        template_to_descriptor(
+            ParameterTemplate("dummy", bounds=DummyBounds())
+        )
+
+
+def test_dict_behavior():
+    entries = {
+        "density": RealDescriptor("density", lower_bound=0, upper_bound=100, units="g/cm^3"),
+        "pressure": RealDescriptor("pressure", lower_bound=0, upper_bound=10000, units="GPa")
+    }
+
+    v = PlatformVocabulary(entries)
+
+    assert len(v) == 2
+    assert set(v) == {"density", "pressure"}
+    assert v["density"] == entries["density"]
+    assert v["pressure"] == entries["pressure"]
+
+
+def test_from_template(fake_project: Project):
+    """Test that only correct scopes and bounds are loaded from templates."""
+    v = PlatformVocabulary.from_templates(fake_project, scope="my_scope")
+    assert len(v) == 1
+    assert list(v) == ["density"]
+    assert v["density"] == density_desc
```

Comment:
It took me several glances to realize that "volume" doesn't appear because we don't support integer descriptors.

Reply:
Commented for clarity.